packages/run-protocol/src/vpool-xyk-amm/pool.js

```diff
@@ -43,210 +43,257 @@ export const makeAddPool = (
-        assert(isNatValue(secondaryPoolAmount.value), 'Pool Secondary');
-
-        // To calculate liquidity, we'll need to calculate alpha from the primary
-        // token's value before, and the value that will be added to the pool
-        const secondaryRequired = AmountMath.make(
-          secondaryBrand,
-          calcSecondaryRequired(
-            centralIn.value,
-            centralPoolAmount.value,
-            secondaryPoolAmount.value,
-            secondaryIn.value,
-          ),
-        );
+  const addLiquidityActualToState = (
+    state,
+    pool,
+    zcfSeat,
+    secondaryAmount,
+    poolCentralAmount,
+    feeSeat,
+  ) => {
+    const { poolSeat, liquidityBrand, liquidityZcfMint, updater } = state;
+
+    // addLiquidity can't be called until the pool has been created. We
+    // verify that the asset is NAT before creating a pool.
+
+    const liquidityValueOut = calcLiqValueToMint(
+      state.liqTokenSupply,
+      zcfSeat.getStagedAllocation().Central.value,
+      poolCentralAmount.value,
+    );
 
-        // Central was specified precisely so offer must provide enough secondary.
-        assert(
-          AmountMath.isGTE(secondaryIn, secondaryRequired),
-          'insufficient Secondary deposited',
-        );
+    const liquidityAmountOut = AmountMath.make(
+      liquidityBrand,
+      liquidityValueOut,
+    );
+    liquidityZcfMint.mintGains(
+      harden({ Liquidity: liquidityAmountOut }),
+      poolSeat,
+    );
+    state.liqTokenSupply += liquidityValueOut;
+
+    poolSeat.incrementBy(
+      zcfSeat.decrementBy(
+        harden({
+          Central: zcfSeat.getStagedAllocation().Central,
+          Secondary: secondaryAmount,
+        }),
+      ),
+    );
 
-        return addLiquidityActual(pool, zcfSeat, secondaryRequired, centralIn);
-      },
-      removeLiquidity: userSeat => {
-        const liquidityIn = userSeat.getAmountAllocated(
-          'Liquidity',
-          liquidityBrand,
-        );
-        const liquidityValueIn = liquidityIn.value;
-        assert(isNatValue(liquidityValueIn), 'User Liquidity');
-        const centralTokenAmountOut = AmountMath.make(
-          centralBrand,
-          calcValueToRemove(
-            liqTokenSupply,
-            pool.getCentralAmount().value,
-            liquidityValueIn,
-          ),
-        );
+    zcfSeat.incrementBy(
+      poolSeat.decrementBy(harden({ Liquidity: liquidityAmountOut })),
+    );
+    if (feeSeat) {
+      zcf.reallocate(poolSeat, zcfSeat, feeSeat);
+    } else {
+      zcf.reallocate(poolSeat, zcfSeat);
+    }
+    zcfSeat.exit();
+    updateUpdaterState(updater, pool);
+    return 'Added liquidity.';
+  };
 
-        const tokenKeywordAmountOut = AmountMath.make(
-          secondaryBrand,
-          calcValueToRemove(
-            liqTokenSupply,
-            pool.getSecondaryAmount().value,
-            liquidityValueIn,
-          ),
-        );
+  const makePool = defineKind(
```

Comment:
is the best documentation on `defineKind` the comment on `defineKindInternal`  in `virtualObjectManager.js`?   The parameter types don't seem to be declared in a way that's visible to VSCode and WebStorm, so I always have to find the docs before I can tell what arguments should be provided in what order.

Reply:
https://github.com/Agoric/agoric-sdk/pull/4603/commits/48dbf0f1fa8b86ee3b6781c4e200fc16cfa854e0#diff-4d09d78ca33d43e2584cfbf5adc00ed1caac51f1edbc10bae3d3c14320e56d30 will help